serializer: remove duplicate of lua serializer

mandesero opened the pull request:
This patch fixes a bug found by the ASAN instrumentation of LuaJIT
allocator [1]. The problem is using a Lua serializer object that has
been cleaned up by GC.

Tarantool build:
 - Tarantool 3.2.0-entrypoint-84-g9d3859b24
 - Target: Linux-x86_64-Debug
 - Compiler: Clang-14.0.0
 - LuaJIT branch: mandesero/dlmalloc-instr
 - CMAKE_FLAGS:
 	```
	-DENABLE_WERROR=ON -DENABLE_ASAN=ON -DENABLE_UB_SANITIZER=ON
	-DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION=OFF -DENABLE_FUZZER=ON
	-DTEST_BUILD=ON -DCMAKE_BUILD_TYPE=RelWithDebInfo
	-DFIBER_STACK_SIZE=640Kb -DLUAJIT_ENABLE_GC64=ON
	-DCMAKE_BUILD_TYPE=Debug
	```
 - ASAN_OPTIONS:
 	```
	detect_invalid_pointer_pairs=1:detect_leaks=1:\
	dump_instruction_bytes=1:heap_profile=0:\
	print_suppressions=0:symbolize=1:unmap_shadow_on_exit=1:\
	redzone=32
	```
 - LSAN_OPTIONS:
 	```
	suppressions=tarantool/asan/lsan.supp
	```

The crash occurs when executing `(tarantool) \set output lua`.

Failing tests:
 - ./test/app-luatest/gh_7031_configure_eos_in_lua_console_test.lua
 - ./test/app-tap/console.test.lua
 - ./test/box/push.test.lua
 - ./app-tap/console_lua.test.lua
 - ./app-luatest/varbinary_test.lua

The `serializer_lua` static member is removed from
`src/box/lua/serialize_lua.c` along with `serializer_lua_init`,
because it is not used by any function other than `serializer_lua_init`,
which is not needed now too.

[1]: Issue https://github.com/tarantool/tarantool/issues/10231

Closes https://github.com/tarantool/tarantool/issues/10177 (this issue is a duplicate of https://github.com/tarantool/tarantool/issues/7404)

NO_TEST=rely on existing tests
NO_CHANGELOG=codehealth
NO_DOC= codehealth

Buristan commented:
Looks related to the #7404.

mkokryashkin commented on src/box/lua/serialize_lua.c:
```diff
@@ -1038,32 +1038,10 @@ lua_parse_opts(lua_State *L, lua_dumper_opts_t *opts)
-	 * won't be unloaded.
-	 */
-	lua_setfield(L, -2, "formatter_lua");
+	serializer_lua = serializer;
```
Yaml serializer above is different from the lua serializer, because it is a static member specific for the console module. Lua serializer is set as a static member in the console module and the same serializer is set as a static member in the lua serializer module. Moreover, lua serializer module is now not responsible for linking the serializer object anywhere, so it static member may hold a pointer to a dead object if someone outside forgets to create a reference.

That static member also seems redundant, because it is not used by any function other than setter, so let's remove it along with the setter.

mkokryashkin reviewed (changes requested):
Please specify more details in the commit message: explanation of the issue, failing test name, build parameters. Also it is important to specify that test fails under ASAN. 

Add tag "Related to #7404"

mkokryashkin reviewed (changes requested):
> All crashes are the result of attempts to access `serializer_lua`
fields in `src/box/lua/console.c`, which was overwritten when
`src/box/lua/serialize_lua.c:lua_serializer_init` was called due
to repeated call of `lua_setfield(L, -2, “formatter_lua”)`.
Typo: s/to repeated/to the repeated/

>So let's remove the static lua serializer member from
`src/box/lua/serialize_lua.c` along with `serializer_lua_init`,
because it is not used by any function other than `serializer_lua_init`.

It is better to paraphrase like the following:
```
The `serializer_lua` static member is removed from
`src/box/lua/serialize_lua.c` along with `serializer_lua_init`,
because it is not used by any function other than `serializer_lua_init`,
which is not needed now too.
```

Otherwise, patch LGTM

ligurio reviewed (commented):
Maksim, thanks for the patch!

1. Actually I don't get why we should describe tarantool version, lsan and asan options in commit message. I believe this information should be added to the ticket #10231 and removed from commit message/
2. reason in `NO_CHANGELOG` and `NO_DOC` can be replaced to `codehealth`. It is not a bugfix, actually.
3. Ilya Verbin said about similar problem (#7404). It is not clear: whether it is a duplicate or not.

mandesero commented:
> Maksim, thanks for the patch!
> 
> 1. Actually I don't get why we should describe tarantool version, lsan and asan options in commit message. I believe this information should be added to the ticket [ASAN & LuaJIT allocator #10231](https://github.com/tarantool/tarantool/issues/10231) and removed from commit message/
> 2. reason in `NO_CHANGELOG` and `NO_DOC` can be replaced to `codehealth`. It is not a bugfix, actually.
> 3. Ilya Verbin said about similar problem ([Unallocated `serializer_lua` used by `lbox_console_format_lua` #7404](https://github.com/tarantool/tarantool/issues/7404)). It is not clear: whether it is a duplicate or not.

Sergey, thanks for your comment! I took the following points into consideration.

1. I removed the description of the tarantool build from the commit message, a link to the issue where the build is described was added to ASAN & LuaJIT allocator #10231.
2. NO_DOC and NO_CHANGELOG were changed to the suggested ones.
3. The current issue is indeed a duplicate ([Unallocated `serializer_lua` used by `lbox_console_format_lua` #7404](https://github.com/tarantool/tarantool/issues/7404)), only found using ASAN. I indicated information about the duplicate in the commit message.

ligurio reviewed (approved):
LGTM,  thanks

Totktonada commented:
Pushed to `master`, a future `3.2.0` release.

Backported to `release/3.1`, a future `3.1.2` release, in PR #10394.

Backported to `release/2.11`, a future `2.11.4` release, in PR #10395.
